From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: sf2000-linux <sf2000-linux@local>
Date: Wed, 13 May 2026 00:00:00 +0000
Subject: [PATCH] recognize mips uclinux targets

Buildroot uses mipsel-buildroot-uclinux-uclibc when BR2_mipsel is
combined with BR2_BINFMT_FLAT.  GCC has MIPS GNU/Linux support and
generic uClinux defaults, but no target-specific MIPS uClinux case, so
configure rejects the tuple before the compiler can be built.

Route mips*-*-uclinux* through the existing MIPS GNU/Linux setup and add
flat.h to mark the target as FLAT-capable.  Add a small MIPS uClinux
header so static FLAT links use crtbegin.o instead of the GNU/Linux
crtbeginT.o object that this target does not build.
---
 gcc/config.gcc            | 26 ++++++++++++++++++++++++++
 gcc/config/mips/uclinux.h | 46 +++++++++++++++++++++++++++++++++++++++
 libgcc/config.host        | 11 +++++++++++
 3 files changed, 83 insertions(+)

diff --git a/gcc/config.gcc b/gcc/config.gcc
index fa96fd74152..02d93f815e0 100644
--- a/gcc/config.gcc
+++ b/gcc/config.gcc
@@ -2707,4 +2707,30 @@ mips*-mti-linux*)
 	gnu_ld=yes
 	gas=yes
 	;;
+mips*-*-uclinux*)			# uClinux MIPS, either endian.
+	tm_file="elfos.h gnu-user.h linux.h linux-android.h glibc-stdint.h flat.h ${tm_file} mips/gnu-user.h mips/linux.h mips/linux-common.h mips/uclinux.h"
+	extra_options="${extra_options} linux.opt linux-android.opt"
+	tm_p_file="${tm_p_file} linux-protos.h"
+	tmake_file="${tmake_file} t-slibgcc t-linux t-glibc"
+	extra_objs="${extra_objs} linux.o"
+	c_target_objs="${c_target_objs} glibc-c.o"
+	cxx_target_objs="${cxx_target_objs} glibc-c.o"
+	d_target_objs="${d_target_objs} glibc-d.o"
+	tm_defines="${tm_defines} ANDROID_DEFAULT=0"
+	target_has_targetcm=yes
+	target_has_targetdm=yes
+	case ${target} in
+		mipsisa32r6*)
+			default_mips_arch=mips32r6
+			;;
+		mipsisa32r2*)
+			default_mips_arch=mips32r2
+			;;
+		mipsisa32*)
+			default_mips_arch=mips32
+			;;
+	esac
+	gnu_ld=yes
+	gas=yes
+	;;
 mips*-*-linux*)				# Linux MIPS, either endian.
diff --git a/gcc/config/mips/uclinux.h b/gcc/config/mips/uclinux.h
new file mode 100644
index 00000000000..b5ee492f2a9
--- /dev/null
+++ b/gcc/config/mips/uclinux.h
@@ -0,0 +1,46 @@
+/* Definitions for MIPS uClinux FLAT targets.
+   Copyright (C) 2026 Free Software Foundation, Inc.
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with GCC; see the file COPYING3.  If not see
+<http://www.gnu.org/licenses/>.  */
+
+#undef TARGET_OS_CPP_BUILTINS
+#define TARGET_OS_CPP_BUILTINS()		\
+  do						\
+    {						\
+      GNU_USER_TARGET_OS_CPP_BUILTINS ();	\
+      builtin_define ("__uClinux__");		\
+    }						\
+  while (0)
+
+/* FLAT executables run at a fixed load address and cannot use the normal
+   GNU/Linux MIPS abicalls/GOT model.  Keep the target non-PIC by default. */
+#undef TARGET_DEFAULT
+#define TARGET_DEFAULT 0
+
+#undef SUBTARGET_ASM_SPEC
+#define SUBTARGET_ASM_SPEC ""
+
+#undef STARTFILE_SPEC
+#define STARTFILE_SPEC \
+  "%{!shared:crt1.o%s} crti.o%s %{shared|pie:crtbeginS.o%s;:crtbegin.o%s}"
+
+#undef ENDFILE_SPEC
+#define ENDFILE_SPEC "crtend.o%s crtn.o%s"
+
+#undef LINK_SPEC
+#define LINK_SPEC GNU_USER_TARGET_LINK_SPEC " -e __start"
+
+#undef LIB_SPEC
+#define LIB_SPEC "%{pthread:-lpthread} -lc"
diff --git a/libgcc/config.host b/libgcc/config.host
index 16a755b510d..0cd98fef242 100644
--- a/libgcc/config.host
+++ b/libgcc/config.host
@@ -1020,6 +1020,17 @@ mips*-*-netbsd*)			# NetBSD/mips, either endian.
 	;;
+mips*-*-uclinux*)			# uClinux MIPS, either endian.
+	extra_parts="$extra_parts crtfastmath.o"
+	tmake_file="${tmake_file} t-crtfm"
+	case ${host} in
+	  mips64r5900* | mipsr5900*)
+	    ;;
+	  *)
+	    tmake_file="${tmake_file} mips/t-mips16 t-slibgcc-libgcc"
+	    ;;
+	esac
+	;;
 mips*-*-linux*)				# Linux MIPS, either endian.
 	extra_parts="$extra_parts crtfastmath.o"
 	tmake_file="${tmake_file} t-crtfm"
 	case ${host} in
 	  mips64r5900* | mipsr5900*)
-- 
2.50.0
